A wrapper for second order tensors to reuse eigenvalue decomposition

hugary1995 opened the pull request:
A wrapper for RankTwoTensor is added to handle matrix exponentiation and more.

close #20375

<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Design
<!--A concise description (design) of the enhancement.-->
A wrapper (tentatively called `FactorizedSymmetricRankTwoTensor`) is added to perform common operations on a RankTwoTensor that requires a spectral decomposition. The wrapper is designed such that a `FactorizedSymmetricRankTwoTensor` is _always_ factorized. const and non-const getters are provided to retrieve the underlying `RankTwoTensor`. Only operations/methods that are able to reuse the spectral factorization are implemented for the wrapper, which prevents potential misuse.

The methods that reuse the existing eigen values and eigen vectors of a matrix (that are implemented in this PR) are
  - rotated
  - transpose
  - operator=
  - operator*=, operator/=
  - operator*, operator/
  - inverse
  - addIa
  - trace
  - det
  - log
  - exp
  - pow
  - sqrt
  - cbrt

I could add more methods if needed.

hugary1995 committed:
FactorizedSymmetricRankTwoTensor

A wrapper for RankTwoTensor is added to handle matrix exponentiation and more.

close #20375

hugary1995 commented on modules/tensor_mechanics/src/materials/ComputeFiniteStrain.C:
```diff
@@ -219,27 +221,10 @@ ComputeFiniteStrain::computeQpIncrements(RankTwoTensor & total_strain_increment,
-
-      rotation_increment = _Fhat[_qp] * invUhat;
-
-      total_strain_increment =
-          N1 * std::log(lambda1) + N2 * std::log(lambda2) + N3 * std::log(lambda3);
+      FactorizedSymmetricRankTwoTensor Chat = _Fhat[_qp].transpose() * _Fhat[_qp];
+      FactorizedSymmetricRankTwoTensor Uhat = Chat.sqrt();
+      rotation_increment = _Fhat[_qp] * Uhat.inverse().get();
+      total_strain_increment = Uhat.log().get();
```
As a demonstration, we could rewrite this piece of code using just four lines.

With the new code, we still only perform spectral decomposition just once. We even saved an iterative `inverse()` from before!

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/964244/) on ccc7875 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20376/site)

This comment will be updated on new commits.

hugary1995 commented:
I would like to use the named constructor `vectorSelfOuterProduct` from #18998 to improve the performance a little bit.

hugary1995 commented on modules/tensor_mechanics/src/materials/ComputeFiniteStrain.C:
```diff
@@ -219,27 +221,10 @@ ComputeFiniteStrain::computeQpIncrements(RankTwoTensor & total_strain_increment,
-
-      rotation_increment = _Fhat[_qp] * invUhat;
-
-      total_strain_increment =
-          N1 * std::log(lambda1) + N2 * std::log(lambda2) + N3 * std::log(lambda3);
+      FactorizedSymmetricRankTwoTensor Chat = _Fhat[_qp].transpose() * _Fhat[_qp];
+      FactorizedSymmetricRankTwoTensor Uhat = MathUtils::sqrt(Chat);
+      rotation_increment = _Fhat[_qp] * Uhat.inverse().get();
+      total_strain_increment = MathUtils::log(Uhat).get();
       break;
     }
```
As a demonstration, we could rewrite this piece of code using just four lines.

With the new code, we still only perform spectral decomposition just once. We even saved an iterative inverse() from before!

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/964528/) on 50e79eb wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>cd0fbf</th>
      <th colspan=4>#20376 50e79e</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.34%</td>
      <td>82.39%</td>
      <td>+0.05%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>73848</td>
      <td>73952</td>
      <td>+104</td>
      <td>75</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15840</td>
      <td>15811</td>
      <td>-29</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20376/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20376/coverage/framework)


## Modules coverage

### Tensor mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>cd0fbf</th>
      <th colspan=4>#20376 50e79e</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.44%</td>
      <td>89.44%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>26169</td>
      <td>26160</td>
      <td>-9</td>
      <td>4</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3090</td>
      <td>3090</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20376/coverage_diff/tensor_mechanics.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20376/coverage/tensor_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20376/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20376/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20376/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20376/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20376/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20376/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20376/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20376/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20376/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20376/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20376/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20376/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20376/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20376/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20376/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20376/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20376/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20376/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20376/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20376/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20376/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20376/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/20376/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20376/coverage/xfem)

</details>



This comment will be updated on new commits.

dschwen commented:
> I would like to use the named constructor `vectorSelfOuterProduct` from #18998 to improve the performance a little bit.

Yeah, of only somebody would review #18998.... *cough* @bwspenc *cough*

dschwen commented on framework/include/utils/FactorizedSymmetricRankTwoTensor.h:
```diff
@@ -0,0 +1,300 @@
+  }
+
+  // @{ getters
+  RankTwoTensorTempl<T> get() const { return assemble(); }
```
I think it would be forward looking to template on the tensor type itself (that could be a template template argument, and then another argument for the scalar type. Or better yet, we add a `value_type` member to `RankTwoTensorTempl<T>` that resolves to `T`. 

That way we can use the wrapper if that SymmetricRankTwoTensor PR ever gets merged. In that context I'd suggest calling the instantiation of you class `FactorizedRankTwoTensor` to distinguish it from a future `FactorizedSymmetricRankTwoTensor`, that is based on `SymmetricRankTwoTensor`.

dschwen replied:
Actually `value_type` already exists (and is inherited from `TypeTensor<T>`)

hugary1995 replied:
Agreed on the template.

In fact this wrapper only works on symmetric second order tensors. As you've probably already noticed, I am manually checking if the input second order tensor is symmetric. So although I may instantiate to `FactorizedRankTwoTensor`, it still really is a `FactorizedSymmetricRankTwoTensor` at its core, and so it may be a little misleading to name if `FactorizedRankTwoTensor`. I am with that though: we could name it `FactorizedRankTwoTensor`, and say it's the users responsibility to make sure that the tensor is symmetric.

hugary1995 replied:
Another possibility is that I only instantiate a `FactorizedSymmetricRankTwoTensor` on `SymmetricRankTwoTensor`, and rely on the implicit conversion from `RankTwoTensor` to `SymmetricRankTwoTensor` at construction time. When the `RankTwoTensor` isn't symmetric, the implicit conversion fails.

I don't remember if your `SymmetricRankTwoTensor` stuff provides such implicit conversion though.

dschwen replied:
It does not. I made those constructors `explicit`

```
  /// Copy constructor from TensorValue<T>
  explicit SymmetricRankTwoTensorTempl(const TensorValue<T> & a);

  /// Copy constructor from TypeTensor<T>
  explicit SymmetricRankTwoTensorTempl(const TypeTensor<T> & a);
```

hugary1995 replied:
Yeah, it's dangerous if not being explicit.

hugary1995 replied:
I'll do the explicit conversion in the constructor of this wrapper then. The check for symmetry is cheap compared to Eigen decomposition.

dschwen replied:
> In fact this wrapper only works on symmetric second order tensors. As you've probably already noticed, I am manually checking if the input second order tensor is symmetric

If you use isSymmetric() for that check it works be a noop for the symmetric class and would still work for the regular class.

hugary1995 replied:
It's templated on tensor type now.

hugary1995 replied:
Wait a sec. I am getting compile errors in dbg saying `RankTwoTensorTempl` doesn't have a method `isSymmetric()`. Is it added in your Mandel notation PR?

hugary1995 replied:
I also added the isSymmetric method.

dschwen commented on framework/include/utils/FactorizedSymmetricRankTwoTensor.h:
```diff
@@ -0,0 +1,300 @@
+{
+  const auto & eigvals = A.eigvals();
+  return FactorizedSymmetricRankTwoTensorTempl<typename CompareTypes<T, T2>::supertype>(
+      {std::pow(eigvals[0], p), std::pow(eigvals[1], p), std::pow(eigvals[2], p)}, A.eigvecs());
```
Is this only working for Internet powers? Or what does the scalar traits stuff mean?

hugary1995 replied:
Fixed.

dschwen commented on framework/include/utils/FactorizedSymmetricRankTwoTensor.h:
```diff
@@ -0,0 +1,300 @@
+  RankTwoTensorTempl<T> you = A.get();
+  for (auto i : make_range(LIBMESH_DIM))
+    for (auto j : make_range(LIBMESH_DIM))
+      if (!MooseUtils::absoluteFuzzyEqual(me(i, j), you(i, j)))
```
Why not return me == you?

hugary1995 replied:
Fixed

dschwen commented:
Let's keep this open. Gary are you planning to come back to it?

hugary1995 commented:
Just leaving a note here: I'll come back to this after the Mandel notation PR is merged as I want to experiment on the templating. I think this could also be an opportunity to re-think the overall design -- it may be better to create a new trait rather than creating `FactorizedXXXTensor` for each and every tensor class.

hugary1995 commented:
I will also need derivative of an operator map of a factorized R2T with respect to the R2T itself in #21167. I'll add those derivatives there.

loganharbour commented:
You should look into `if constexpr()` type checks instead of `enable_if` madness

hugary1995 commented:
Good point. I was very excited about c++17 but completely forgot about it when I actually write code :smile:

hugary1995 commented:
I'll just put the highlight here.

This is a piece of code in `ComputeFiniteStrain.C` before this PR:
```cpp
    case DecompMethod::EigenSolution:
    {
      std::vector<Real> e_value(3);
      RankTwoTensor e_vector, N1, N2, N3;

      RankTwoTensor Chat = _Fhat[_qp].transpose() * _Fhat[_qp];
      Chat.symmetricEigenvaluesEigenvectors(e_value, e_vector);

      const Real lambda1 = std::sqrt(e_value[0]);
      const Real lambda2 = std::sqrt(e_value[1]);
      const Real lambda3 = std::sqrt(e_value[2]);

      N1.vectorOuterProduct(e_vector.column(0), e_vector.column(0));
      N2.vectorOuterProduct(e_vector.column(1), e_vector.column(1));
      N3.vectorOuterProduct(e_vector.column(2), e_vector.column(2));

      const RankTwoTensor Uhat = N1 * lambda1 + N2 * lambda2 + N3 * lambda3;
      const RankTwoTensor invUhat(Uhat.inverse());

      rotation_increment = _Fhat[_qp] * invUhat;

      total_strain_increment =
          N1 * std::log(lambda1) + N2 * std::log(lambda2) + N3 * std::log(lambda3);
      break;
    }
```

with this PR, this piece of code can be rewritten in the following more concise and efficient form:
```cpp
    case DecompMethod::EigenSolution:
    {
      FactorizedRankTwoTensor Chat = _Fhat[_qp].transpose() * _Fhat[_qp];
      FactorizedRankTwoTensor Uhat = MathUtils::sqrt(Chat);
      rotation_increment = _Fhat[_qp] * Uhat.inverse().get();
      total_strain_increment = MathUtils::log(Uhat).get();
      break;
    }
```

I expect similar improvements in plasticity, crystal plasticity, and the lagrangian mechanics materials.

hugary1995 commented:
@dschwen Do you have any more comments?

reverendbedford commented on unit/src/FactorizedRankTwoTensorTest.C:
```diff
@@ -0,0 +1,177 @@
+  EXPECT_FALSE(Af1 == Bf);
+}
+
+TEST(FactorizedRankTwoTensor, inverse)
```
I guess not then...

reverendbedford commented on framework/src/utils/FactorizedRankTwoTensor.C:
```diff
@@ -0,0 +1,102 @@
+  return me == you;
+}
+
+template <typename T>
```
Oh right, symmetric.

hugary1995 replied:
Yeah, this is using the symmetry assumption.

reverendbedford reviewed (commented):
I don't see anything wrong in the PR.  But maybe a couple of general comments:

1. Really it should be `FactorizedSymmetricRankTwoTensor`
2. I should be non-lazy and check the derivative formula.  Or is there an easy way to check that in the unit tests with finite differencing?
3. Is is worthwhile to add `det`?  It's easy to calculate once you have the decomposition.

hugary1995 commented:
> 1. Really it should be `FactorizedSymmetricRankTwoTensor`

That's the name I had originally. But the issue is that we would want to template on the wrapped tensor type. Right now I instantiate
```cpp
typedef FactorizedRankTwoTensorTempl<RankTwoTensor> FactorizedRankTwoTensor;
```
so that I could distinguish it with a future
```cpp
typedef FactorizedSymmetricRankTwoTensorTempl<SymmetricRankTwoTensor> FactorizedSymmetricRankTwoTensor;
```

Speaking of which, I should add that instantiation since we now have that `SymmetricRankTwoTensor`.

> 2. I should be non-lazy and check the derivative formula.  Or is there an easy way to check that in the unit tests with finite differencing?

Yeah, I should think of a way to test that.

> 3. Is is worthwhile to add `det`?  It's easy to calculate once you have the decomposition.

Yep

reverendbedford commented:
I see your point on naming, particularly if in the future someone expands this to take a general `RankTwoTensor` as input and provide the non-symmetric eigendecomposition.  However, the way you have things now no matter what they're putting in they are getting a symmetric factorization out.  And you're only checking they put a symmetric `RankTwoTensor` in in debug mode.

You almost want two inputs instead of one:
1. The tensor to factorize (`RankTwoTensor` or `SymmetricRankTwoTensor`)
2. The method to use to factorize.  Right now just `SpectralDecomposition` (= symmetric eigendecomposition) but others could be possible in the future.

reverendbedford commented:
That was the alternative I was going to suggest (just always check), so that's fine with me for what it's worth.

hugary1995 committed:
Instantiate FactorizedSymmetricRankTwoTensor

hugary1995 committed:
Add trace() and det(); Add missing implementations for two explicit constructors of SymmetricRankTwoTensor; Add isSymmetric(); Always check if the given tensor is symmetric when constructing a FactorizedRankTwoTensor.

hugary1995 committed:
Add unit tests for derivatives

hugary1995 commented:
Okay, so I have added
- instantiations for `[AD]SymmetricRankTwoTensor` and their usage demonstration in `ADComputeFiniteStrain`. To do so I added the missing explicit constructors in those classes
- `FactorizedRankTwoTensorTempl<T>::trace()` and `FactorizedRankTwoTensorTempl<T>::det()`
- unit tests comparing analytical derivatives to finite difference derivatives, which indeed uncovered a bug (now fixed).

moosebuild commented:
Job [Test debug](https://civet.inl.gov/job/1082617/) on b46e466 : invalidated by @hugary1995

dschwen commented on framework/include/utils/FactorizedRankTwoTensor.h:
```diff
@@ -0,0 +1,291 @@
+    return static_cast<T2>(assemble());
+  }
+  const std::vector<typename T::value_type> & eigvals() const { return _eigvals; }
+  std::vector<typename T::value_type> eigvals() { return _eigvals; }
```
Uhm, is that non const getter even needed? why would we get them by copy in that case? why not also get them by const reference?

dschwen commented on framework/include/utils/SymmetricRankTwoTensor.h:
```diff
@@ -285,6 +288,9 @@ class SymmetricRankTwoTensorTempl
   template <typename T2>
   bool operator==(const SymmetricRankTwoTensorTempl<T2> & a) const;
 
+  /// Test for symmetry. Surprisingly this is always true.
```
:-D

dschwen commented:
Ok, this is neat. The getter comment can be addressed later (unless there is a reason the code is what it is)

loganharbour commented:
This broke the world of unit testing: https://civet.inl.gov/event/93415/

hugary1995 commented:
I'll patch it and tag you. Should be easy to fix. Sorry about that.

dschwen commented:
Ok, how exactly did this happen with all tests being green on each PR? (this is the same issue that popped up in the conflict resolution in thw other PR)

dschwen commented:
@loganharbour this https://civet.inl.gov/event/93430/ _should_ pass now

loganharbour commented:
It was merged after the tests hadn't run in 10 days; the build env changed

dschwen commented:
Ah, right of course. That has bitten me before...

hugary1995 commented:
Good, everything back to green after a costco shopping. Daniel did the right fix -- the unit test has been fixed in another PR of mine which Daniel just merged.
